polar-core/src/vm.rs

```diff
@@ -1406,9 +1406,34 @@ impl PolarVirtualMachine {
             Value::Expression(_) => {
                 return self.query_for_operation(&term);
             }
+            Value::Variable(_a_symbol) => {
+                let val = self.deref(term);
+
+                if val == *term {
+                    // variable was unbound
+                    // apply a constraint to variable that it must be truthy
+                    self.push_goal(Goal::Query {
+                        term: term.clone_with_value(value!(op!(Eq, term.clone(), term!(true)))),
+                    })?;
+                } else {
+                    self.push_goal(Goal::Query { term: val })?;
```

Comment:
This is going to make the `x = f(1) and x` case "work", i.e. query `f(1)`.

We spoke about making predicates _not_ first class objects, so perhaps the place to handle this is instead the unification?

Reply:
When you have cycles you, I, and Alex should revisit all this.

Reply:
I don't necessarily need to be in a full discussion about this -- just adding an input. We generally try to have querying for unbound, constrained or bound values mean the same thing. Isa does not abide by this right now and it causes a bunch of issues.

As written, querying for an unbound results in `x == true`. Querying for a bound value may result in `x == true` if the value is bound to a boolean, or something else if the value is bound to another value.  

If we proceeded with this:

`x and x = f(x)`: What does this mean? `x` will add `x == true`. If the grounder works correctly with calls (it probably doesn't) this query would fail.
`x = f(x) and x` has a different result, we query for `f(x)`.

Having different results if the conjunction is written in a different order is not ideal.